agent crashed when networkpolicystats feature was enabled and an antrea networkpolicy with drop action was created

**Describe the bug**
After I enabled NetworkPolicyStats feature and created an Antrea NetworkPolicy with drop action, all agents crashed:
```
panic: runtime error: index out of range [6] with length 6

goroutine 252 [running]:
github.com/vmware-tanzu/antrea/pkg/agent/openflow.parseDropFlow(0xc00121826e, 0x5c, 0x1e22dd4, 0x4, 0x30, 0xc000fa3880)
        /antrea/pkg/agent/openflow/network_policy.go:1408 +0x2d9
github.com/vmware-tanzu/antrea/pkg/agent/openflow.parseMetricFlow(0xc00121826e, 0x5c, 0x1e2b1e7, 0xa, 0xffffffffffffffff, 0x0)
        /antrea/pkg/agent/openflow/network_policy.go:1435 +0x6e
github.com/vmware-tanzu/antrea/pkg/agent/openflow.(*client).NetworkPolicyMetrics.func1(0xc00052b380, 0x4, 0x4)
        /antrea/pkg/agent/openflow/network_policy.go:1451 +0xe7
github.com/vmware-tanzu/antrea/pkg/agent/openflow.(*client).NetworkPolicyMetrics(0xc0002eb600, 0x40f5b0)
        /antrea/pkg/agent/openflow/network_policy.go:1467 +0x147
github.com/vmware-tanzu/antrea/pkg/agent/stats.(*Collector).collect(0xc000525360, 0xc0001c4000)
        /antrea/pkg/agent/stats/collector.go:105 +0x4f
github.com/vmware-tanzu/antrea/pkg/agent/stats.(*Collector).Run(0xc000525360, 0xc0005296e0)
        /antrea/pkg/agent/stats/collector.go:83 +0xe7
created by main.run
        /antrea/cmd/antrea-agent/agent.go:258 +0x2385
```

I checked the openflow of metrics flows:
```
# ovs-ofctl dump-flows br-int table=101
 cookie=0x6050000000000, duration=30711.157s, table=101, n_packets=1, n_bytes=54, priority=200,ct_state=+new,ct_label=0x2/0xffffffff,ip actions=resubmit(,105)
 cookie=0x6050000000000, duration=30711.157s, table=101, n_packets=0, n_bytes=0, priority=200,ct_state=-new,ct_label=0x2/0xffffffff,ip actions=resubmit(,105)
 cookie=0x6050000000000, duration=25822.004s, table=101, n_packets=0, n_bytes=0, priority=200,reg0=0x100000/0x100000,reg3=0x1 actions=drop
```
And the code parsing the flows: https://github.com/vmware-tanzu/antrea/blob/521b177430326d6eb129f9c85743031f3e1bd903/pkg/agent/openflow/network_policy.go#L1398-L1409

It seemed the function expected `reg3=` to be the 6th seg (starting from 0) while the flow has only 6 segs.
I guess it was broken after merging ipv6 branch which removes `ip` match from the metrics flows to support both IPv4 and IPv6. 

**To Reproduce**
1. Enable networkpolicystats
2. Create an Antrea NetworkPolicy with drop action.

**Expected**
1. Agent shouldn't crash.
2. Add e2e test for anp with drop action to prevent it from being broken another time.

**Actual behavior**
1. Agent crashed
2. It was not caught by any test

**Versions:**
Please provide the following information:
 - Antrea version (Docker image tag). 0.11.1

--- a/pkg/agent/openflow/network_policy.go
+++ b/pkg/agent/openflow/network_policy.go
@@ -1395,33 +1395,29 @@ func (c *client) ReassignFlowPriorities(updates map[uint16]uint16, table binding
 	return nil
 }
 
-func parseDropFlow(flow string) (uint32, types.RuleMetric) {
-	// example format:
-	// table=101, n_packets=9, n_bytes=666, priority=200,ip,reg0=0x100000/0x100000,reg3=0x5 actions=drop
-	segs := strings.Split(flow, ",")
+func parseDropFlow(flowMap map[string]string) (uint32, types.RuleMetric) {
 	m := types.RuleMetric{}
-	pkts, _ := strconv.ParseUint(segs[1][strings.Index(segs[1], "=")+1:], 10, 64)
+	pkts, _ := strconv.ParseUint(flowMap["n_packets"], 10, 64)
 	m.Packets = pkts
 	m.Sessions = pkts
-	bytes, _ := strconv.ParseUint(segs[2][strings.Index(segs[2], "=")+1:], 10, 64)
+	bytes, _ := strconv.ParseUint(flowMap["n_bytes"], 10, 64)
 	m.Bytes = bytes
-	id, _ := strconv.ParseUint(segs[6][strings.Index(segs[6], "0x")+2:strings.Index(segs[6], " ")], 16, 64)
+	reg3 := flowMap["reg3"]
+	id, _ := strconv.ParseUint(reg3[:strings.Index(reg3, " ")], 0, 64)
 	return uint32(id), m
 }
 
-func parseAllowFlow(flow string) (uint32, types.RuleMetric) {
-	// example format:
-	// table=101, n_packets=0, n_bytes=0, priority=200,ct_state=-new,ct_label=0x1/0xffffffff,ip actions=goto_table:105
-	segs := strings.Split(flow, ",")
+func parseAllowFlow(flowMap map[string]string) (uint32, types.RuleMetric) {
 	m := types.RuleMetric{}
-	pkts, _ := strconv.ParseUint(segs[1][strings.Index(segs[1], "=")+1:], 10, 64)
+	pkts, _ := strconv.ParseUint(flowMap["n_packets"], 10, 64)
 	m.Packets = pkts
-	if strings.Contains(segs[4], "+") { // ct_state=+new
+	if strings.Contains(flowMap["ct_state"], "+") { // ct_state=+new
 		m.Sessions = pkts
 	}
-	bytes, _ := strconv.ParseUint(segs[2][strings.Index(segs[2], "=")+1:], 10, 64)
+	bytes, _ := strconv.ParseUint(flowMap["n_bytes"], 10, 64)
 	m.Bytes = bytes
-	idRaw := segs[5][strings.Index(segs[5], "0x")+2 : strings.Index(segs[5], "/")]
+	ct_label := flowMap["ct_label"]
+	idRaw := ct_label[strings.Index(ct_label, "0x")+2 : strings.Index(ct_label, "/")]
 	if len(idRaw) > 8 { // only 32 bits are valid.
 		idRaw = idRaw[:len(idRaw)-8]
 	}
@@ -1429,12 +1425,27 @@ func parseAllowFlow(flow string) (uint32, types.RuleMetric) {
 	return uint32(id), m
 }
 
+func parseFlowToMap(flow string) map[string]string {
+	split := strings.Split(flow, ",")
+	flowMap := make(map[string]string)
+	for _, seg := range split {
+		equalIndex := strings.Index(seg, "=")
+		flowMap[strings.TrimSpace(seg[:equalIndex])] = strings.TrimSpace(seg[equalIndex+1:])
+	}
+	return flowMap
+}
+
 func parseMetricFlow(flow string) (uint32, types.RuleMetric) {
 	dropIdentifier := "reg0"
-	if strings.Contains(flow, dropIdentifier) {
-		return parseDropFlow(flow)
+	flowMap := parseFlowToMap(flow)
+	// example allow flow format:
+	// table=101, n_packets=0, n_bytes=0, priority=200,ct_state=-new,ct_label=0x1/0xffffffff,ip actions=goto_table:105
+	// example drop flow format:
+	// table=101, n_packets=9, n_bytes=666, priority=200,reg0=0x100000/0x100000,reg3=0x5 actions=drop
+	if _, ok := flowMap[dropIdentifier]; ok {
+		return parseDropFlow(flowMap)
 	} else {
-		return parseAllowFlow(flow)
+		return parseAllowFlow(flowMap)
 	}
 }
 

--- a/pkg/agent/openflow/network_policy_test.go
+++ b/pkg/agent/openflow/network_policy_test.go
@@ -769,7 +769,7 @@ func TestParseMetricFlow(t *testing.T) {
 		metric types.RuleMetric
 	}{
 		"Drop flow": {
-			flow: "table=101, n_packets=9, n_bytes=666, priority=200,ip,reg0=0x100000/0x100000,reg3=0x5 actions=drop",
+			flow: "table=101, n_packets=9, n_bytes=666, priority=200,reg0=0x100000/0x100000,reg3=0x5 actions=drop",
 			rule: 5,
 			metric: types.RuleMetric{
 				Bytes:    666,

--- a/test/e2e/antreapolicy_test.go
+++ b/test/e2e/antreapolicy_test.go
@@ -18,6 +18,7 @@ import (
 	"context"
 	"fmt"
 	"strings"
+	"sync"
 	"testing"
 	"time"
 
@@ -26,6 +27,7 @@ import (
 	v1 "k8s.io/api/core/v1"
 	v1net "k8s.io/api/networking/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
 	"k8s.io/apimachinery/pkg/util/wait"
 
 	secv1alpha1 "github.com/vmware-tanzu/antrea/pkg/apis/security/v1alpha1"
@@ -1219,4 +1221,118 @@ func TestAntreaPolicyStatus(t *testing.T) {
 		return anp.Status == expectedStatus, nil
 	})
 	assert.NoError(t, err, "Antrea ClusterNetworkPolicy failed to reach expected status")
+}
+
+// TestANPNetworkPolicyStatsWithDropAction tests antreanetworkpolicystats can correctly collect dropped packets stats from ANP if
+// networkpolicystats feature is enabled
+func TestANPNetworkPolicyStatsWithDropAction(t *testing.T) {
+	data, err := setupTest(t)
+	if err != nil {
+		t.Fatalf("Error when setting up test: %v", err)
+	}
+	defer teardownTest(t, data)
+	skipIfAntreaPolicyDisabled(t, data)
+
+	if err := testData.mutateAntreaConfigMap(func(data map[string]string) {
+		antreaControllerConf, _ := data["antrea-controller.conf"]
+		antreaControllerConf = strings.Replace(antreaControllerConf, "#  NetworkPolicyStats: false", "  NetworkPolicyStats: true", 1)
+		data["antrea-controller.conf"] = antreaControllerConf
+		antreaAgentConf, _ := data["antrea-agent.conf"]
+		antreaAgentConf = strings.Replace(antreaAgentConf, "#  NetworkPolicyStats: false", "  NetworkPolicyStats: true", 1)
+		data["antrea-agent.conf"] = antreaAgentConf
+	}, true, true); err != nil {
+		t.Fatalf("Failed to enable NetworkPolicyStats feature: %v", err)
+	}
+
+	serverName, serverIPs, cleanupFunc := createAndWaitForPod(t, data, data.createNginxPodOnNode, "test-server-", "")
+	defer cleanupFunc()
+
+	clientName, _, cleanupFunc := createAndWaitForPod(t, data, data.createBusyboxPodOnNode, "test-client-", "")
+	defer cleanupFunc()
+	k8sUtils, err = NewKubernetesUtils(data)
+	failOnError(err, t)
+	p10 := float64(10)
+	intstr80 := intstr.FromInt(80)
+	dropAction := secv1alpha1.RuleActionDrop
+	selectorB := metav1.LabelSelector{MatchLabels: map[string]string{"antrea-e2e": clientName}}
+	selectorC := metav1.LabelSelector{MatchLabels: map[string]string{"antrea-e2e": serverName}}
+	protocol := v1.ProtocolUDP
+
+	var anp = &secv1alpha1.NetworkPolicy{
+		ObjectMeta: metav1.ObjectMeta{Namespace: testNamespace, Name: "np1", Labels: map[string]string{"antrea-e2e": "np1"}},
+		Spec: secv1alpha1.NetworkPolicySpec{
+			AppliedTo: []secv1alpha1.NetworkPolicyPeer{
+				{PodSelector: &selectorC},
+			},
+			Priority: p10,
+			Ingress: []secv1alpha1.Rule{
+				{
+					Ports: []secv1alpha1.NetworkPolicyPort{
+						{
+							Port:     &intstr80,
+							Protocol: &protocol,
+						},
+					},
+					From: []secv1alpha1.NetworkPolicyPeer{
+						{
+							PodSelector: &selectorB,
+						},
+					},
+					Action: &dropAction,
+				},
+			},
+			Egress: []secv1alpha1.Rule{},
+		},
+	}
+
+	if _, err = k8sUtils.CreateOrUpdateANP(anp); err != nil {
+		failOnError(fmt.Errorf("create ANP failed for ANP %s: %v", anp.Name, err), t)
+	}
+
+	// Wait for a few seconds in case that connections are established before policies are enforced.
+	time.Sleep(networkPolicyDelay)
+
+	sessionsPerAddressFamily := 10
+	var wg sync.WaitGroup
+	for i := 0; i < sessionsPerAddressFamily; i++ {
+		wg.Add(1)
+		go func() {
+			if clusterInfo.podV4NetworkCIDR != "" {
+				cmd := []string{"/bin/sh", "-c", fmt.Sprintf("echo test | nc -w 4 -u %s 80", serverIPs.ipv4.String())}
+				data.runCommandFromPod(testNamespace, clientName, busyboxContainerName, cmd)
+			}
+			if clusterInfo.podV6NetworkCIDR != "" {
+				cmd := []string{"/bin/sh", "-c", fmt.Sprintf("echo test | nc -w 4 -u %s 80", serverIPs.ipv6.String())}
+				data.runCommandFromPod(testNamespace, clientName, busyboxContainerName, cmd)
+			}
+			wg.Done()
+		}()
+	}
+	wg.Wait()
+
+	totalSessions := 0
+	if clusterInfo.podV4NetworkCIDR != "" {
+		totalSessions += sessionsPerAddressFamily
+	}
+	if clusterInfo.podV6NetworkCIDR != "" {
+		totalSessions += sessionsPerAddressFamily
+	}
+
+	if err := wait.Poll(5*time.Second, defaultTimeout, func() (bool, error) {
+		stats, err := data.crdClient.StatsV1alpha1().AntreaNetworkPolicyStats(testNamespace).Get(context.TODO(), "np1", metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		t.Logf("Got AntreaNetworkPolicy stats: %v", stats)
+		if stats.TrafficStats.Sessions != int64(totalSessions) {
+			return false, nil
+		}
+		if stats.TrafficStats.Packets < stats.TrafficStats.Sessions || stats.TrafficStats.Bytes < stats.TrafficStats.Sessions {
+			return false, fmt.Errorf("Neither 'Packets' nor 'Bytes' should be smaller than 'Sessions'")
+		}
+		return true, nil
+	}); err != nil {
+		failOnError(err, t)
+	}
+	k8sUtils.Cleanup(namespaces)
 }

--- a/test/e2e/infra/vagrant/playbook/roles/common/tasks/docker.yml
+++ b/test/e2e/infra/vagrant/playbook/roles/common/tasks/docker.yml
@@ -33,7 +33,7 @@
     groups: docker
     append: yes
     create_home: yes
-  when: ansible_version.full >= "2.5.0"
+  when: ansible_version.full is version_compare('2.5.0', '>=')
 
 - name: Add test user to docker group (Ansible < 2.5.0)
   user:
@@ -41,7 +41,7 @@
     groups: docker
     append: yes
     createhome: yes
-  when: ansible_version.full < "2.5.0"
+  when: ansible_version.full is version_compare('2.5.0', '<')
 
 # Install docker-py so that docker images can be pulled
 - pip:

--- a/test/e2e/networkpolicy_test.go
+++ b/test/e2e/networkpolicy_test.go
@@ -105,9 +105,9 @@ func TestNetworkPolicyStats(t *testing.T) {
 	// Wait for a few seconds in case that connections are established before policies are enforced.
 	time.Sleep(2 * time.Second)
 
-	sessions := 10
+	sessionsPerAddressFamily := 10
 	var wg sync.WaitGroup
-	for i := 0; i < sessions; i++ {
+	for i := 0; i < sessionsPerAddressFamily; i++ {
 		wg.Add(1)
 		go func() {
 			if clusterInfo.podV4NetworkCIDR != "" {
@@ -123,25 +123,25 @@ func TestNetworkPolicyStats(t *testing.T) {
 	}
 	wg.Wait()
 
-	multiple := 0
+	totalSessions := 0
 	if clusterInfo.podV4NetworkCIDR != "" {
-		multiple++
+		totalSessions += sessionsPerAddressFamily
 	}
 	if clusterInfo.podV6NetworkCIDR != "" {
-		multiple++
+		totalSessions += sessionsPerAddressFamily
 	}
-	totalSessions := sessions * multiple
+
 	if err := wait.Poll(5*time.Second, defaultTimeout, func() (bool, error) {
 		for _, np := range []string{"test-networkpolicy-ingress", "test-networkpolicy-egress"} {
-			metric, err := data.crdClient.StatsV1alpha1().NetworkPolicyStats(testNamespace).Get(context.TODO(), np, metav1.GetOptions{})
+			stats, err := data.crdClient.StatsV1alpha1().NetworkPolicyStats(testNamespace).Get(context.TODO(), np, metav1.GetOptions{})
 			if err != nil {
 				return false, err
 			}
-			t.Logf("Got NetworkPolicy metric: %v", metric)
-			if metric.TrafficStats.Sessions != int64(totalSessions) {
+			t.Logf("Got NetworkPolicy stats: %v", stats)
+			if stats.TrafficStats.Sessions != int64(totalSessions) {
 				return false, nil
 			}
-			if metric.TrafficStats.Packets < metric.TrafficStats.Sessions || metric.TrafficStats.Bytes < metric.TrafficStats.Sessions {
+			if stats.TrafficStats.Packets < stats.TrafficStats.Sessions || stats.TrafficStats.Bytes < stats.TrafficStats.Sessions {
 				return false, fmt.Errorf("Neither 'Packets' nor 'Bytes' should be smaller than 'Sessions'")
 			}
 		}

